File: server/src/services/group-member/group-member.service.ts
import { Injectable, ForbiddenException, NotFoundException, BadRequestException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { GroupMember } from 'src/models/group-member.entity';
import { Group } from 'src/models/group.entity';
import { User } from 'src/models/user.entity';

@Injectable()
export class GroupMembersService {
    constructor(
        @InjectRepository(GroupMember) private groupMemberRepo: Repository<GroupMember>,
        @InjectRepository(Group) private groupRepo: Repository<Group>,
        @InjectRepository(User) private userRepo: Repository<User>,
    ) { }

    async addMember(groupId: string, userId: string, requesterId: string) {
        const group = await this.groupRepo.findOne({
            where: { id: groupId },
            relations: ['createdBy'],
        });
        if (!group) throw new NotFoundException('Group not found');

        if (group.createdBy.id !== requesterId) {
            throw new ForbiddenException('Only the group creator can add members');
        }

        const user = await this.userRepo.findOne({ where: { id: userId } });
        if (!user) throw new NotFoundException('User not found');

        const member = this.groupMemberRepo.create({ group, user });
        return this.groupMemberRepo.save(member);
    }

    async removeMember(groupId: string, userId: string, requesterId: string) {
        const group = await this.groupRepo.findOne({
            where: { id: groupId },
            relations: ['createdBy'],
        });
        if (!group) throw new NotFoundException('Group not found');

        if (group.createdBy.id !== requesterId) {
            throw new ForbiddenException('Only the group creator can remove members');
        }

        await this.groupMemberRepo.delete({ group: { id: groupId }, user: { id: userId } });
        return { message: 'Member removed' };
    }

    async listMembers(groupId: string) {
        return this.groupMemberRepo.find({
            where: { group: { id: groupId } },
            relations: ['user'],
        });
    }

    async leaveGroup(groupId: string, userId: string) {
        const group = await this.groupRepo.findOne({
            where: { id: groupId },
            relations: ['createdBy'], // 👈 ensure we can access createdBy.id
        });

        if (!group) {
            throw new NotFoundException('Group not found');
        }

        // ✅ Prevent creator from leaving
        if (group.createdBy.id === userId) {
            throw new BadRequestException('Group creator cannot leave the group');
        }

        const result = await this.groupMemberRepo.delete({
            group: { id: groupId },
            user: { id: userId },
        });

        if (result.affected === 0) {
            throw new NotFoundException('You are not a member of this group');
        }

        return { message: 'You have left the group' };
    }


}

File: server-java/moneyjar/moneyjar/src/main/java/com/tarungattu/moneyjar/services/group/GroupMemberService.java
package com.tarungattu.moneyjar.services.group;

import com.tarungattu.moneyjar.models.Group;
import com.tarungattu.moneyjar.models.GroupMember;
import com.tarungattu.moneyjar.models.User;
import com.tarungattu.moneyjar.repository.GroupMemberRepository;
import com.tarungattu.moneyjar.repository.GroupRepository;
import com.tarungattu.moneyjar.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
public class GroupMemberService {

    @Autowired
    private GroupMemberRepository groupMemberRepository;

    @Autowired
    private GroupRepository groupRepository;

    @Autowired
    private UserRepository userRepository;

    @Transactional
    public GroupMember addMember(UUID groupId, UUID userId, UUID requesterId) {
        Group group = groupRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));

        if (!group.getCreatedBy().getId().equals(requesterId)) {
            throw new RuntimeException("Only the group creator can add members");
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Check if already a member
        boolean isMember = group.getMembers().stream()
                .anyMatch(m -> m.getUser().getId().equals(userId));
        if (isMember) {
            throw new RuntimeException("User is already a member of this group");
        }

        GroupMember member = new GroupMember();
        member.setGroup(group);
        member.setUser(user);

        return groupMemberRepository.save(member);
    }

    @Transactional
    public void removeMember(UUID groupId, UUID userId, UUID requesterId) {
        Group group = groupRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));

        if (!group.getCreatedBy().getId().equals(requesterId)) {
            throw new RuntimeException("Only the group creator can remove members");
        }

        if (group.getCreatedBy().getId().equals(userId)) {
             throw new RuntimeException("Cannot remove the group creator");
        }

        GroupMember member = groupMemberRepository.findByGroupIdAndUserId(groupId, userId)
                .orElseThrow(() -> new RuntimeException("Member not found"));

        groupMemberRepository.delete(member);
    }

    @Transactional
    public void leaveGroup(UUID groupId, UUID userId) {
        Group group = groupRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));

        if (group.getCreatedBy().getId().equals(userId)) {
            throw new RuntimeException("Group creator cannot leave the group");
        }

        GroupMember member = groupMemberRepository.findByGroupIdAndUserId(groupId, userId)
                .orElseThrow(() -> new RuntimeException("You are not a member of this group"));

        groupMemberRepository.delete(member);
    }
}
